### PR TITLE
Add 'recovery export'.

### PR DESCRIPTION
**Pull request by floitsch**

### CODE REVIEW

File: src/cli/fleet.toit [hunk 1]
```diff
@@ -471,6 +472,11 @@ class Fleet:
     new-file := fleet-file_.with --recovery-urls=[]
     new-file.write
 
+  recovery-info -> ByteArray:
+    broker.server-config.fill-certificate-ders: certificate-roots.MAP[it].raw
+    json-config := broker.server-config.to-json --der-serializer=: base64.encode it
```
Comment (kasperl) [inline src/cli/fleet.toit:477]: This is not the format the device expects.

File: src/cli/cmds/fleet.toit [hunk 2]
```diff
@@ -953,6 +971,9 @@ migration-stop parsed/cli.Parsed config/Config cache/Cache ui/Ui:
       joined := quoted.join ", "
       ui.info "Stopped migration for broker(s) $joined."
 
+recovery-file-name fleet-id/uuid.Uuid -> string:
+  return "recover-$(fleet-id).json"
```
[the remark this answers is not in the document]
Comment (floitsch) [reply src/cli/cmds/fleet.toit]: done.

File: src/cli/fleet.toit (hunk 1 above)
Comment (floitsch) [reply src/cli/fleet.toit]: done.

File: src/shared/server-config.toit [hunk 3]
```diff
@@ -220,7 +232,10 @@ class ServerConfigHttp extends ServerConfig:
     if root-certificate-names:
       result["root_certificate_names"] = root-certificate-names
     if root-certificate-ders:
-      result["root_certificate_ders"] = root-certificate-ders.map: der-serializer.call it
+      if base64:
+        result["root_certificate_ders64"] = root-certificate-ders.map: base64-lib.encode it
```
Comment (kasperl) [inline src/shared/server-config.toit:236]: It isn't super clear to me why we need this section?

File: src/cli/fleet.toit [hunk 4]
```diff
@@ -471,6 +472,14 @@ class Fleet:
     new-file := fleet-file_.with --recovery-urls=[]
     new-file.write
 
+  recovery-info -> ByteArray:
+    broker.server-config.fill-certificate-ders: certificate-roots.MAP[it].raw
+    certificate-der/ByteArray? := null
+    json-config := broker.server-config.to-json
```
Comment (kasperl) [inline src/cli/fleet.toit:478]: Can we end up with something that has a ders/ders64 section here? Hmm. That wouldn't work for the device side.

File: src/cli/fleet.toit [hunk 5]
```diff
@@ -471,6 +472,14 @@ class Fleet:
     new-file := fleet-file_.with --recovery-urls=[]
     new-file.write
 
+  recovery-info -> ByteArray:
+    broker.server-config.fill-certificate-ders: certificate-roots.MAP[it].raw
+    certificate-der/ByteArray? := null
```
Comment (kasperl) [inline src/cli/fleet.toit:477]: Unused?

File: src/cli/cmds/fleet.toit [hunk 6]
```diff
@@ -1002,3 +1027,29 @@ recovery-list parsed/cli.Parsed config/Config cache/Cache ui/Ui:
...
+  with-devices-fleet parsed config cache ui: | fleet/FleetWithDevices |
+    recovery-info := fleet.recovery-info
+    path := fs.join directory (recovery-file-name --fleet-id=fleet.id)
+    file.write-content --path=path recovery-info
+    ui.info "Exported recovery information to '$path'."
+    recovery-urls := fleet.recovery-urls
+    full-urls := recovery-urls.map: | url/string |
+      "$url/$(recovery-file-name --fleet-id=fleet.id)"
+    if not recovery-urls.is-empty:
+      ui.info "Devices with the current recovery-servers configuration will try to"
```
Comment (kasperl) [inline src/cli/cmds/fleet.toit:1043]:
```suggestion
      ui.info "Devices with the current recovery servers configuration will try to"
```

File: src/shared/server-config.toit (hunk 3 above)
Comment (kasperl) [reply src/shared/server-config.toit:236]: I get it now.

**Review by kasperl — APPROVED** [review]
Looks good to me.

File: src/cli/fleet.toit (hunk 5 above)
Comment (floitsch) [reply src/cli/fleet.toit]: done.

**Pull request merged**